parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java

```diff
@@ -0,0 +1,364 @@
+  }
+
+  private ByteBuffer ensure(ByteBuffer buffer, int size) {
+    if (buffer == null) {
+      buffer = allocator.allocate(size);
+    } else if (buffer.capacity() >= size) {
+      buffer.clear();
+    } else {
+      allocator.release(buffer);
+      release(buffer);
+      buffer = allocator.allocate(size);
+    }
+    return buffer;
+  }
+
+  ByteBuffer release(ByteBuffer buffer) {
+    if (buffer != null) {
+      allocator.release(buffer);
+    }
+    return null;
+  }
+
+  @Override
+  protected BytesCompressor createCompressor(final CompressionCodecName codecName, final CompressionCodec codec,
+                                             int pageSize) {
+
+    if (codec == null) {
+      return new NoopCompressor();
+    } else if (codecName == CompressionCodecName.SNAPPY) {
+      // avoid using the Parquet Snappy codec since it allocates direct buffers at awkward spots.
+      return new SnappyCompressor();
+    } else {
+
+      // todo: move zlib above since it also generates allocateDirect calls.
+      return new HeapBytesCompressor(codecName, codec, pageSize);
+    }
+  }
+
+  @Override
+  protected DirectBytesDecompressor createDecompressor(final CompressionCodec codec) {
+    // This is here so that debugging can be done if we see inconsistencies between our decompression and upstream
+    // decompression.
+    // if (true) {
+    // return new HeapFakeDirect(codec);
+    // }
+
+    if (codec == null) {
+      return new NoopDecompressor();
+    } else if (DirectCodecPool.INSTANCE.codec(codec).supportsDirectDecompression()) {
+      return new FullDirectDecompressor(codec);
+    } else {
+      return new IndirectDecompressor(codec);
+    }
+  }
+
+  public void close() {
+    release();
+  }
+
+  /**
+   * Keeping this here for future debugging versus using custom implementations below.
+   */
+  private class HeapFakeDirect extends DirectBytesDecompressor {
+
+    private final ExposedHeapBytesDecompressor innerCompressor;
+
+    public HeapFakeDirect(CompressionCodec codec){
+      innerCompressor = new ExposedHeapBytesDecompressor(codec);
+    }
+
+    @Override
+    public void decompress(ByteBuffer input, int compressedSize, ByteBuffer output, int uncompressedSize)
+        throws IOException {
+      BytesInput uncompressed = decompress(new ByteBufBytesInput(input), uncompressedSize);
+      output.clear();
+      output.put(uncompressed.toByteArray(), 0, uncompressedSize);
+    }
+
+    @Override
+    public BytesInput decompress(BytesInput paramBytesInput, int uncompressedSize) throws IOException {
+      return innerCompressor.decompress(paramBytesInput, uncompressedSize);
+    }
+
+    @Override
+    protected void release() {
+      innerCompressor.release();
+    }
+
+  }
+
+  private class ExposedHeapBytesDecompressor extends HeapBytesDecompressor {
+    public ExposedHeapBytesDecompressor(CompressionCodec codec) {
+      super(codec);
+    }
+
+    public void release() {
+      super.release();
+    }
```

Comment:
Changed to protected access in the abstract class above as well as all implementations.